### DISCUSSION TITLE
The scrolling upwards may cause periodic jitter

### DISCUSSION THREAD

**Discussion by mc-public** [krzyzanowskim/STTextView#26] [Questions]
I am currently trying to understand the behavior of `NSTextViewportLayoutControllerDelegate` in `STTextView`. I have noticed that when copying a large amount of text into the text view (about 30,000 lines) and then dragging the scroll bar to the bottom and scrolling upwards with the trackpad at a constant speed, there is a periodic jittering effect. Interestingly, if I quickly browse through the area to be scrolled first and then scroll upwards within that area, this phenomenon disappears. I have tried modifying the content in `STTextView+NSTextViewportLayoutControllerDelegate.swift` but have not been able to determine a solution. Is this issue happening within TextKit2 internally?
<img width="503" alt="Example" src="https://github.com/krzyzanowskim/STTextView/assets/106484439/67994e45-b237-4418-9f17-78a710968d4b">

**Comment by krzyzanowskim** [comment]
Although nothing is documented, my current hypothesis is that it's the TextKit2 bug. Looks like layout estimate sizes even though the report it as not estimated. Later on, the size is corrected, but damage already has been done.

> [at times, reported height is way above the final value for the document it result in "jumping" the scroller, as frame height grow and shrink](https://github.com/krzyzanowskim/STTextView/blob/da53620e55a175d55bf20d8bb693e332a7f9e353/Sources/STTextView/STTextView.swift#L914)

**Reply by mc-public** [reply]
I carefully reviewed Apple's [TextKit2 sample project](https://developer.apple.com/documentation/uikit/textkit/using_textkit_2_to_interact_with_text). In the demo app of this project, even for very large documents, there is no periodic jitter while scrolling. After a thorough analysis, I found that Apple uses animations to avoid this issue. 
![pic](https://github.com/krzyzanowskim/STTextView/assets/106484439/9f12a467-50e0-4ed4-9288-f16f82a962c0)
After removing the animations from these CALayers, the periodic jitter reappeared.
